weaknesses/MASVS-STORAGE/MASWE-0005.md

```diff
@@ -1,46 +1,45 @@
+- **Bypass Protection Mechanism**: Hardcoded API keys can make it easier to bypass app protection mechanisms. Attackers can use this to access restricted content, cheat in app functionality, or unlock features that are intended for purchase, impacting both revenue and user experience.
 
 ## Modes of Introduction
 
-Sensitive data can be hardcoded in several areas:
+API keys can be hardcoded in several areas:
 
 - **App Source Code**: directly embedded in the app source code.
 - **App Assets**: included in files that are destined for the final deliverable app package (typically APK/IPA), such as configuration files, manifest files, and resource files.
 - **Libraries**: configuration files or source code for third-party, first-party libraries or any other app dependencies.
 
 ## Mitigations
 
-To mitigate the risks associated with hardcoded sensitive data, developers should:
-
-- Use platform specific solutions to encrypt sensitive data, such as Android Keystore or iOS Keychain, before storing them locally.
-- Implement proper key management practices, including key rotation and using environment-specific keys.
-- Avoid storing sensitive data within the app package or source code, and instead securely retrieve it from server-side services that provide secure storage, access control, and auditing for sensitive data. For example, AWS Secrets Manager, Azure Key Vault, or Google Cloud Secret Manager are some popular managed secrets storage solutions. The app can securely retrieve the necessary secrets at runtime through secure, authenticated API calls.
-- Where possible, generate cryptographic keys dynamically on the device, rather than using predefined keys, and ensure that they don't leave the platform-provided keystore. This approach reduces the risk associated with key transmission and storage.
-- If data must be hardcoded, such as some API keys, be sure to configure them with the minimum required permissions to reduce the impact in case of exposure. Many services allow you to create keys with restricted access, which limits the operations that can be performed.
+- Use a stateful API service that provides secure authentication, client validation, and session controls. Implement dynamic tokens that expire after a reasonably short time (e.g., 1 hour). This can help reduce the impact of key exposure. Also, ensure proper error handling and logging to detect and respond to unauthorized access attempts. Consider using OAuth 2.0 and security libraries like AppAuth to simplify secure OAuth flows.
```

Comment:
Maybe we should mention whitebox cryptography here too? Obviously it's not a full mitigation but it raises the bar for the attacker.

Reply:
Good point. Added. Thanks Jan!